src/new/wallet.js

```diff
@@ -0,0 +1,441 @@
+          reject(`Wrong network. server=${info.network} expected=${this.network}`);
+        }
+      }, (error) => {
+        console.log('Version error:', error);
+        this.setState(Wallet.CLOSED);
+        reject(error);
+      });
+    });
+    return promise;
+  }
+
+  /**
+   * Close the connections and stop emitting events.
+   **/
+  stop() {
+    // TODO Double check that we are properly cleaning things up.
+    // See: https://github.com/HathorNetwork/hathor-wallet-headless/pull/1#discussion_r369859701
+    ws.stop()
+    ws.removeListener('is_online', this.onConnectionChange);
+    ws.removeListener('reload_data', this.reloadData);
+    ws.removeListener('addresses_loaded', this.onAddressesLoaded);
+    ws.removeListener('wallet', this.handleWebsocketMsg);
+    this.serverInfo = null;
+    this.setState(Wallet.CLOSED);
+  }
+
+  /**
+   * Returns the balance for each token in tx, if the input/output belongs to this wallet
+   */
+  getTxBalance(tx) {
+    const myKeys = []; // TODO
+    const balance = {};
+    for (const txout of tx.outputs) {
+      if (wallet.isAuthorityOutput(txout)) {
+        continue;
+      }
+      if (txout.decoded && txout.decoded.address
+          && txout.decoded.address in myKeys) {
+        if (!balance[txout.token]) {
+          balance[txout.token] = 0;
+        }
+        balance[txout.token] += txout.value;
+      }
+    }
+
+    for (const txin of tx.inputs) {
+      if (wallet.isAuthorityOutput(txin)) {
+        continue;
+      }
+      if (txin.decoded && txin.decoded.address
+          && txin.decoded.address in myKeys) {
+        if (!balance[txin.token]) {
+          balance[txin.token] = 0;
+        }
+        balance[txin.token] -= txin.value;
+      }
+    }
+
+    return balance;
+  }
+
+  getTokenData() {
+    if (this.tokenUid) {
+      if (this.tokenUid === HATHOR_TOKEN_CONFIG.uid) {
+        // Hathor token we don't get from the full node
+        this.token = HATHOR_TOKEN_CONFIG;
+      } else {
+        // Get token info from full node
+        // XXX This request might take longer than the ws connection to start
+        // so it's possible (but hard to happen) that the wallet will change to
+        // READY state with token still null.
+        // I will keep it like that for now but to protect from this
+        // we should change to READY only after both things finish
+        walletApi.getGeneralTokenInfo(this.tokenUid, (response) => {
+          if (response.success) {
+            this.token = {
+              uid: this.tokenUid,
+              name: response.name,
+              symbol: response.symbol,
+            }
+          } else {
+            throw Error(response.message);
+          }
+        });
+      }
+    }
+  }
+
+  isReady() {
+    return this.state === Wallet.READY;
+  }
+}
+
+// State constants.
+Wallet.CLOSED =  0;
+Wallet.CONNECTING = 1;
+Wallet.SYNCING = 2;
+Wallet.READY = 3;
+
+Wallet.getHumanState = (state) => {
```

Comment:
It should be inside the `class` definition. Even though it seems the same to set it here, it is different. When it is inside the `class`, the method is defined in the prototype. When we put it here, it is out of the prototype.

Reply:
Removed and added to wallet headless